File: bms/go/geo/geo_s2_Cap_IsValid/geo_s2_Cap_IsValid.go
package test

import (
	"math"
)

const (
	geo_s1_StraightChordAngle = geo_s1_ChordAngle(4)
)

func (c geo_s2_Cap) IsValid() bool {
	return c.center.geo_r3_Vector.IsUnit() && c.radius <= geo_s1_StraightChordAngle
}

func (v geo_r3_Vector) IsUnit() bool {
	const epsilon = 5e-14
	return math.Abs(v.Norm2()-1) <= epsilon
}

func (v geo_r3_Vector) Norm2() float64 { return v.Dot(v) }

func (v geo_r3_Vector) Dot(ov geo_r3_Vector) float64 {
	return float64(v.X*ov.X) + float64(v.Y*ov.Y) + float64(v.Z*ov.Z)
}

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

type geo_s1_ChordAngle float64

type geo_s2_typeTag uint32

type geo_s2_WedgeRel int

type geo_s2_CrossingType int

type geo_s2_axis int

type geo_s2_CellRelation int

type geo_s2_ShapeIndexIteratorPos int

type geo_s2_Direction int

type geo_s2_crossingTarget int

type geo_s2_Crossing int

type geo_s2_VertexModel int

type geo_r3_Axis int

type geo_s1_Angle float64

File: translations_only/success/go/geo/geo_s2_Cap_IsValid/claude3/Hinted/geo_s2_Cap_IsValid.rs


use std::f64::consts::PI;

const GEO_S1_STRAIGHT_CHORD_ANGLE: f64 = PI / 90.0;

struct GeoS2Cap {
    center: Box<GeoS2Point>,
    radius: GeoS1ChordAngle,
}

impl GeoS2Cap {
    fn is_valid(&self) -> bool {
        self.center.geo_r3_vector.is_unit() && self.radius <= GEO_S1_STRAIGHT_CHORD_ANGLE
    }
}

struct GeoR3Vector {
    x: f64,
    y: f64,
    z: f64,
}

impl GeoR3Vector {
    fn is_unit(&self) -> bool {
        const EPSILON: f64 = 5e-14;
        (self.norm2() - 1.0).abs() <= EPSILON
    }

    fn norm2(&self) -> f64 {
        self.dot(self)
    }

    fn dot(&self, ov: &GeoR3Vector) -> f64 {
        self.x * ov.x + self.y * ov.y + self.z * ov.z
    }
}

struct GeoS2Point {
    geo_r3_vector: GeoR3Vector,
}

type GeoS1ChordAngle = f64;

#[derive(Clone, Copy)]
struct GeoS2TypeTag(u32);

#[derive(Clone, Copy)]
struct GeoS2WedgeRel(i32);

#[derive(Clone, Copy)]
struct GeoS2CrossingType(i32);

#[derive(Clone, Copy)]
struct GeoS2Axis(i32);

#[derive(Clone, Copy)]
struct GeoS2CellRelation(i32);

#[derive(Clone, Copy)]
struct GeoS2ShapeIndexIteratorPos(i32);

#[derive(Clone, Copy)]
struct GeoS2Direction(i32);

#[derive(Clone, Copy)]
struct GeoS2CrossingTarget(i32);

#[derive(Clone, Copy)]
struct GeoS2Crossing(i32);

#[derive(Clone, Copy)]
struct GeoS2VertexModel(i32);

#[derive(Clone, Copy)]
struct GeoR3Axis(i32);

type GeoS1Angle = f64;

